src/ethereum/frontier/trie.py

```diff
@@ -63,78 +263,85 @@ def nibble_list_to_compact(x: Bytes, terminal: bool) -> bytearray:
 
     return compact
 
 
-def map_keys(
-    obj: Mapping[Bytes, Node], secured: bool = True
-) -> Mapping[Bytes, Node]:
+def bytes_to_nibble_list(bytes: Bytes) -> Bytes:
+    """
+    Converts a `Bytes` into to a sequence of nibbles (bytes with value < 16).
+
+    Parameters
+    ----------
+    bytes: `Bytes`
+        The `Bytes` to convert.
+
+    Returns
+    -------
+    nibble_list : `Bytes`
+        The `Bytes` in nibble-list format.
     """
-    Maps all compact keys to nibble-list format. Optionally hashes the keys.
+    nibble_list = bytearray(2 * len(bytes))
+    for i in range(2 * len(bytes)):
+        byte_idx = i // 2
+        if i % 2 == 0:
+            # get upper nibble
+            nibble_list[i] = (bytes[byte_idx] & 0xF0) >> 4
+        else:
+            # get lower nibble
+            nibble_list[i] = bytes[byte_idx] & 0x0F
+    return Bytes(nibble_list)
+
+
+def prepare_trie(
+    obj: Mapping[Bytes, Node],
+    secured: bool = True,
+    get_storage_root: Callable[[Bytes], Bytes] = None,
+) -> Mapping[Bytes, Bytes]:
+    """
+    Prepares the trie for root calculation. Removes values that are empty,
+    hashes the keys (if `secured == True`) and encodes all the nodes.
 
     Parameters
     ----------
     obj :
         Underlying trie key-value pairs.
     secured :
-        Denotes whether the keys should be hashed. Defaults to `true`.
+        Denotes whether the keys should be hashed. Defaults to `True`.
+    get_storage_root :
+        Function to get the storage root of an account. Needed to encode
+        `Account` objects.
 
     Returns
     -------
     out : `Mapping[eth1spec.base_types.Bytes, Node]`
         Object with keys mapped to nibble-byte form.
     """
-    mapped: MutableMapping[Bytes, Node] = {}
-
-    # skip empty values, these are defined to be omitted from the trie
-    skip: Set[Bytes] = set()
-    for (k, v) in obj.items():
-        if v == b"":
-            skip.add(k)
+    mapped: MutableMapping[Bytes, Bytes] = {}
 
     for (preimage, value) in obj.items():
-        if preimage in skip:
-            continue
-
-        # "secure" tries hash keys once before construction
-        key = crypto.keccak256(preimage) if secured else preimage
-
-        nibble_list = bytearray(2 * len(key))
-        for i in range(2 * len(key)):
-            byte_idx = i // 2
-            if i % 2 == 0:
-                # get upper nibble
-                nibble_list[i] = (key[byte_idx] & 0xF0) >> 4
-            else:
-                # get lower nibble
-                nibble_list[i] = key[byte_idx] & 0x0F
-
-        mapped[Bytes(nibble_list)] = value
+        if isinstance(value, Account):
+            assert get_storage_root is not None
+            encoded_value = encode_node(value, get_storage_root(preimage))
+        else:
+            encoded_value = encode_node(value)
+        # skip empty values, these are defined to be omitted from the trie
+        if value != b"":
```

Comment:
Can we remove empty values in the modifier functions, and just have an `assert` here?

It would make me uncomfortable to have multiple possible Python representations for the same conceptual trie.